### DISCUSSION TITLE
Unguided Exercise - Express.js + Mongoose - Advanced Virtual Science Lab Exercise: Titration Simulation API - Part 2/3

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#222] [Tasks]
Logical continuation of the unguided exercise #221 with the use of MongoDB usage.

---

In this exercise, we are going to progress our virtual science lab titration simulation to use MongoDB for data persistence, using Mongoose as the Object Data Modeling (ODM) library. You will be creating schemas, setting up CRUD endpoints, and testing the application using Postman.

### MongoDB and Mongoose Setup

Before you begin, make sure MongoDB is installed and running on your system. You can download it from the official MongoDB website. Then, set up a new Node.js project and install the necessary packages:

```bash
npm init -y
npm install express mongoose body-parser
```

### Mongoose Schemas

Create the Mongoose schemas for acids, bases, indicators, and titrations. Place these in a `models` directory in separate files.

#### models/Acid.js
```javascript
const mongoose = require('mongoose');

const acidSchema = new mongoose.Schema({
  name: String,
  formula: String,
  molarMass: Number
});

module.exports = mongoose.model('Acid', acidSchema);
```

#### models/Base.js
```javascript
const mongoose = require('mongoose');

const baseSchema = new mongoose.Schema({
  name: String,
  formula: String,
  molarMass: Number
});

module.exports = mongoose.model('Base', baseSchema);
```

#### models/Indicator.js
```javascript
const mongoose = require('mongoose');

const indicatorSchema = new mongoose.Schema({
  name: String,
  transitionpHRange: {
    acidic: Number,
    basic: Number
  },
  colors: {
    acidic: String,
    neutral: String,
    basic: String
  }
});

module.exports = mongoose.model('Indicator', indicatorSchema);
```

#### models/Titration.js
```javascript
const mongoose = require('mongoose');

const titrationSchema = new mongoose.Schema({
  acid: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Acid'
  },
  base: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Base'
  },
  indicator: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Indicator'
  },
  volumeBaseAdded: Number,
  pH: Number,
  observations: [{
    volumeBaseAdded: Number,
    pH: Number
  }]
});

module.exports = mongoose.model('Titration', titrationSchema);
```

### Express Server Setup

In your `server.js` file, set up Express and connect to MongoDB with Mongoose.

#### server.js
```javascript
const express = require('express');
const mongoose = require('mongoose');
const bodyParser = require('body-parser');

const Acid = require('./models/Acid');
const Base = require('./models/Base');
const Indicator = require('./models/Indicator');
const Titration = require('./models/Titration');

const app = express();

app.use(bodyParser.json());

mongoose.connect('mongodb://localhost:27017/titrationLab', {
  useNewUrlParser: true,
  useUnifiedTopology: true
});

// CRUD endpoint setup goes here...

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server is running on port ${PORT}`);
});
```

### CRUD Endpoints

Next, set up the CRUD endpoints. I will provide you with the structure and comments to guide you through the implementation. You should implement the logic within each route.

#### CRUD for Acids

- GET `/acids` to retrieve all acids.
- POST `/acids` to create a new acid.
- GET `/acids/:id` to retrieve a specific acid by ID.
- PUT `/acids/:id` to update a specific acid by ID.
- DELETE `/acids/:id` to delete a specific acid by ID.

#### CRUD for Bases

- Implement similar endpoints for bases as for acids.

#### CRUD for Indicators

- Implement similar endpoints for indicators as for acids.

#### CRUD for Titrations

- GET `/titrations` to retrieve all titrations with referenced data populated.
- POST `/titrations` to create a new titration.
- GET `/titrations/:id` to retrieve a specific titration by ID.
- PUT `/titrations/:id` to add an observation to a titration.
- DELETE `/titrations/:id` to delete a specific titration by ID.

### Sample Data

You’ll need to create a script that populates your database with sample data. You can use the data structure provided earlier or create your own.

### Testing with Postman

Create a Postman collection for your endpoints. For each endpoint, define the following:

- The HTTP method and URL.
- The expected request body if applicable (for POST and PUT requests).
- The expected response body.
- Test scripts to validate the response.

Save your requests in a Postman collection and use it to test your server's endpoints, ensuring they correctly interact with the MongoDB database.

### Integration with Frontend

After the backend is tested and functioning as expected, it

 can be integrated with a frontend using d3.js or three.js for data visualization and real-time interaction. The frontend would make API calls to the backend to retrieve data and render it in the browser, enabling students to visualize and understand the titration process dynamically.

This exercise challenges you to apply your knowledge of Node.js, Express, and MongoDB with Mongoose, as well as testing APIs with Postman. You'll also get a chance to see how backend development ties into frontend visualization tools, offering a full-stack development experience.

---

Now, let's continue by filling out the CRUD endpoint logic in your `server.js` file and setting up the initial data population script.

#### server.js (continued)

```javascript
// ... previous server setup code ...

// CRUD endpoints for Titrations
// GET all Titrations
app.get('/titrations', async (req, res) => {
  try {
    const titrations = await Titration.find().populate('acid base indicator');
    res.json(titrations);
  } catch (error) {
    res.status(500).send(error);
  }
});

// POST a new Titration
app.post('/titrations', async (req, res) => {
  try {
    const titration = new Titration(req.body);
    const savedTitration = await titration.save();
    res.status(201).json(savedTitration);
  } catch (error) {
    res.status(400).send(error);
  }
});

// GET a single Titration by ID
app.get('/titrations/:id', async (req, res) => {
  try {
    const titration = await Titration.findById(req.params.id).populate('acid base indicator');
    if (!titration) res.status(404).send('Titration not found');
    else res.json(titration);
  } catch (error) {
    res.status(500).send(error);
  }
});

// PUT to add an observation to a Titration
app.put('/titrations/:id', async (req, res) => {
  try {
    const titration = await Titration.findByIdAndUpdate(
      req.params.id,
      { $push: { observations: req.body } },
      { new: true, runValidators: true }
    );
    if (!titration) res.status(404).send('Titration not found');
    else res.json(titration);
  } catch (error) {
    res.status(400).send(error);
  }
});

// DELETE a specific Titration by ID
app.delete('/titrations/:id', async (req, res) => {
  try {
    const titration = await Titration.findByIdAndDelete(req.params.id);
    if (!titration) res.status(404).send('Titration not found');
    else res.status(200).send('Titration deleted');
  } catch (error) {
    res.status(500).send(error);
  }
});

// ... Add CRUD for acids, bases, and indicators similar to above ...
```

### Initial Data Population Script

You might want to create a script that can be run to populate your MongoDB with initial data for testing.

#### populateDB.js

```javascript
const mongoose = require('mongoose');
const Acid = require('./models/Acid');
const Base = require('./models/Base');
const Indicator = require('./models/Indicator');
const Titration = require('./models/Titration');

const data = {
  acids: [
    { name: 'Hydrochloric Acid', formula: 'HCl', molarMass: 36.46 },
    // ... more acids
  ],
  bases: [
    { name: 'Sodium Hydroxide', formula: 'NaOH', molarMass: 40.00 },
    // ... more bases
  ],
  indicators: [
    { name: 'Phenolphthalein', transitionpHRange: { acidic: 8.2, basic: 10.0 }, colors: { acidic: 'colorless', neutral: 'faint pink', basic: 'fuchsia' } },
    // ... more indicators
  ]
  // No titrations here as they would be created through the API
};

mongoose.connect('mongodb://localhost:27017/titrationLab', {
  useNewUrlParser: true,
  useUnifiedTopology: true
}, async () => {
  console.log('Connected to MongoDB');

  await Acid.deleteMany({});
  await Base.deleteMany({});
  await Indicator.deleteMany({});
  await Titration.deleteMany({});

  await Acid.insertMany(data.acids);
  await Base.insertMany(data.bases);
  await Indicator.insertMany(data.indicators);

  console.log('Database has been populated!');
  process.exit(0);
});
```

Run the script using Node.js to populate your database:
```bash
node populateDB.js
```

### Postman Testing

With your server running, use Postman to create a new collection for your API. For each type of request, you'll set up:

- **URL**: The endpoint's URL.
- **Method**: The HTTP method (GET, POST, PUT, DELETE).
- **Headers**: Content-Type set to `application/json`.
- **Body**: For POST and PUT requests, a raw JSON payload.
- **Tests**: You can write JavaScript tests within Postman to validate the response automatically.

Example for POST `/titrations`:
- **URL**: `http://localhost:300

0/titrations`
- **Method**: POST
- **Body**:
```json
{
  "acid": "<Acid ObjectId>",
  "base": "<Base ObjectId>",
  "indicator": "<Indicator ObjectId>",
  "volumeBaseAdded": 0,
  "pH": 7
}
```

After setting up all the requests in Postman, you can run the collection and verify that all the endpoints are working as expected.

### Integration with Frontend

Once the backend is complete and you have verified that all the endpoints work, you can proceed to connect it with a frontend. The frontend would handle user inputs and visualization, while the backend would provide the data and handle the logic.

### Wrapping Up

This exercise takes you through the full development cycle of a server with a MongoDB database, from setting up the schemas and server logic to populating the database and testing the API with Postman. By the end of it, you should have a solid understanding of creating and managing a backend system that can later be connected with various frontend frameworks and libraries for a complete application experience.

**Comment by UmarAikawa** [comment]
Dear Sir, I have created databases, collections using mongosh and now I can't find them via the terminal. Please how can I retrieve them? Could it be that they are located somewhere in my computer? Where are they being stored?

Thanks as you respond.
